Add readme for validations (#49723)
woocommerce/woocommerce: packages/js/product-editor/src/contexts/validation-context/README.md

# Validations and error handling

This directory contains some components and hooks used for validations in the product editor.

See [Product Editor Development Handbook](../../../../../../docs/product-editor-development/README.md)
for more information.

> ## Add a link to error snackbar

Reviewer: I think we could remove this sub-heading, it comes across a bit confusing.

> ### What happens when there is an error in a field?

Reviewer: ```suggestion
### What happens when there is an error in the form?
```

> 1. Every field [will be validated](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L87-L110).

Reviewer: Is this true? I don't think this is strictly the case right?
To my understanding only the fields that make use of `useValidation` or `useValidations` hooks get registered and will get validated?

So maybe we can rephrase this:
`Fields registered in the validator context will get validated. A field can be registered by making use of the useValidation hook`
Maybe show a quick example as a sub-bullet.

Author: Done in [be38da9](https://github.com/woocommerce/woocommerce/pull/49723/commits/be38da928a6429fd51eed853c2fe3eebe15e30f9)

> 2. An object with the [fields with errors](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L74) will be returned.

Reviewer: ```suggestion
2. An object consisting of the error/validation message, the context, and the validatorId will be returned ([link](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L74) ).
The `context` contains the block Id, and the `validatorId` a unique ID for the validator specifically ( generally a prefix with the block id ).
```
What does this object look like, I see the **step 3** has an example. In this case I think step 3 should be a sub-bullet.

3. If, for instance, the name field is empty, the validation will fail and will throw an object like this:

    ```javascript

>     { message: 'Product name is required.'; context: [block id]; validatorId: [prefix + block id] }

Reviewer: Moving this as a sub-bullet would be helpful

Author: Done in [be38da9](https://github.com/woocommerce/woocommerce/pull/49723/commits/be38da928a6429fd51eed853c2fe3eebe15e30f9)

    ```

> 4. This is the result of the [name validator](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/blocks/product-fields/name/edit.tsx#L84-L101) and the [validatorId addition](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L69).

Reviewer: This should also be a sub-bullet of step 2, but instead of linking the code we may want to list it as an example:
```
async function nameValidator() {
			if ( ! name || name === AUTO_DRAFT_NAME ) {
				return {
					message: __( 'Product name is required.', 'woocommerce' ),
					context: clientId,
				};
			}

			if ( name.length > 120 ) {
				return {
					message: __(
						'Please enter a product name shorter than 120 characters.',
						'woocommerce'
					),
					context: clientId,
				};
			}
		},
```

Author: Done in [be38da9](https://github.com/woocommerce/woocommerce/pull/49723/commits/be38da928a6429fd51eed853c2fe3eebe15e30f9)

> 5. Then, [the props needed](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/components/header/publish-button/publish-button.tsx#L65-L71) to show the error snackbar will be returned.

Reviewer: Hmm I think we should also re-phrase this. Maybe something like this:

```
3. If an error is present on the form we will show an error snackbar with the error message. We will actually include a **show error** link if the field with the relevant error is not visible ( like on another tab ). Clicking the **show error** link will direct users to the relevant field.
     - We create this link by making use of the `context` property ( the block id ), this makes use of the `useBlocksHelper()` hook to get the parent tab id. We can do this by making use of the `core/block-editor` store and using `getBlockParentsByBlockName` ( link to relevant code )
```

Author: Done in [be38da9](https://github.com/woocommerce/woocommerce/pull/49723/commits/be38da928a6429fd51eed853c2fe3eebe15e30f9)

    - When the field with [the error is not visible](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L105), a link pointing to it will be added to the snackbar.
    - Otherwise, the error will be dismissed automatically.

> 6. The hook `useErrorHandler` is used to get the [error props](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L79).

Reviewer: This could be a sub-bullet of the above step

Author: Done in [be38da9](https://github.com/woocommerce/woocommerce/pull/49723/commits/be38da928a6429fd51eed853c2fe3eebe15e30f9)

    - The error shown will depend [on the error code](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L92).
    - [As you can see here](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L157-L162), if the error doesn't have a code, the default message will be `Failed to save product.`
    - The context is used to [get the parent tab](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-blocks-helper/use-blocks-helper.ts#L7) id and the validatorId to [focus on the field](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L68).

> Finally, [the snackbar with the error](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/components/header/publish-button/publish-button.tsx#L70) message and props will be displayed.

Reviewer: Could we show a screenshot or a GIF here?

Author: Done in [be38da9](https://github.com/woocommerce/woocommerce/pull/49723/commits/be38da928a6429fd51eed853c2fe3eebe15e30f9) and [23fe8cd](https://github.com/woocommerce/woocommerce/pull/49723/commits/23fe8cd0dfe5548fc1e8b305844d42f04dd05d4c)

# Validations and error handling

This directory contains some components and hooks used for validations in the product editor.

> ## Add a link to error snackbar

Reviewer: Could we still remove this heading?

### What happens when there is an error in the form?

> 1- Fields registered in the validator context [will get validated](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L87-L110). A field can be registered by making use of the useValidation hook.

Reviewer: ```suggestion
1. Fields registered in the validator context [will get validated](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L87-L110). A field can be registered by making use of the useValidation hook.
```
Formatting is off otherwise

> -   For instance:

Reviewer: ```suggestion
     - For instance:
```

Reviewer: Sub-bullet should be five spaces indented.

```javascript
const {
	ref: myRef,
	error: myValidationError,
	validate: validateMyField,
} = useValidation <
Product >
( 'myfield',
async function myFieldValidator() {
	if ( ! myField ) {
		return {
			message: 'My error message',
			context: clientId,
		};
	}
},
[ myField ] );
```

> 2- An object consisting of the error/validation message, the context, and the validatorId will be returned ([link](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L74) ).

Reviewer: ```suggestion
2. If a field has an error, it returns an object consisting of the error/validation message, the context, and the validatorId ([link](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L74) ).
```

-   The `context` contains the block Id, and the `validatorId` a unique ID for the validator specifically ( generally a prefix with the block id ).
-   If, for instance, the name field is empty, the validation will fail and will throw an object like this:

    ```javascript
    { message: 'Product name is required.'; context: [block id]; validatorId: [prefix + block id] }
    ```

-   This is the result of the name validator and the [validatorId addition](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/contexts/validation-context/validation-provider.tsx#L69).

    ```javascript
    async function nameValidator() {
      if ( ! name || name === AUTO_DRAFT_NAME ) {
        return {
          message: __( 'Product name is required.', 'woocommerce' ),
          context: clientId,
        };
      }

      if ( name.length > 120 ) {
        return {
          message: __(
            'Please enter a product name shorter than 120 characters.',
            'woocommerce'
          ),
          context: clientId,
        };
      }
    },
    ```

> 3- If an error is present on the form we will show an error snackbar with the error message. We will actually include a **show error** link if the field with the relevant error is not visible ( like on another tab ). Clicking the **show error** link will direct users to the relevant field.

Reviewer: ```suggestion
3. If an error is present on the form we will show an error snackbar with the error message. We will actually include a **View error** link if the field with the relevant error is not visible ( like on another tab ). Clicking the **View error** link will direct users to the relevant field.
```

Sorry I suggested **Show error** but it was **View error**

>     - We create this link by making use of the `context` property (the block id), this makes use of the `useBlocksHelper()` hook to get the parent tab id. We can do this by making use of the `core/block-editor` store and using `getBlockParentsByBlockName` (link to relevant code).

Reviewer: I think adding the dot in the numbered bullet above should fix this weird formatting where its in a code block.
As it would be nice if it was just a normal sub-bullet list.

    - The hook `useErrorHandler` is used to get the [error props](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L79).
    - The error shown will depend [on the error code](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L92).
    - [As you can see here](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L157-L162), if the error doesn't have a code, the default message will be `Failed to save product.`
    - The context is used to [get the parent tab](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-blocks-helper/use-blocks-helper.ts#L7) id and the validatorId to [focus on the field](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/hooks/use-error-handler.ts#L68).

Finally, [the snackbar with the error](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/product-editor/src/components/header/publish-button/publish-button.tsx#L70) message and props will be displayed.

> ![Product editor error snackbar](https://developer.woocommerce.com/wp-content/uploads/sites/2/2024/07/product-editor-error-snack-bar-e1721670028482.png)

Reviewer: Nice!!
